cmd/relay/bgs/bgs.go

```diff
@@ -0,0 +1,1375 @@
+
+The firehose WebSocket path is at:  /xrpc/com.atproto.sync.subscribeRepos
+`
+
+func (bgs *BGS) HandleHomeMessage(c echo.Context) error {
+	return c.String(http.StatusOK, homeMessage)
+}
+
+type AuthToken struct {
+	gorm.Model
+	Token string `gorm:"index"`
+}
+
+func (bgs *BGS) lookupAdminToken(tok string) (bool, error) {
+	var at AuthToken
+	if err := bgs.db.Find(&at, "token = ?", tok).Error; err != nil {
+		return false, err
+	}
+
+	if at.ID == 0 {
+		return false, nil
+	}
+
+	return true, nil
+}
+
+func (bgs *BGS) CreateAdminToken(tok string) error {
+	exists, err := bgs.lookupAdminToken(tok)
+	if err != nil {
+		return err
+	}
+
+	if exists {
+		return nil
+	}
+
+	return bgs.db.Create(&AuthToken{
+		Token: tok,
+	}).Error
+}
+
+func (bgs *BGS) checkAdminAuth(next echo.HandlerFunc) echo.HandlerFunc {
+	return func(e echo.Context) error {
+		ctx, span := tracer.Start(e.Request().Context(), "checkAdminAuth")
+		defer span.End()
+
+		e.SetRequest(e.Request().WithContext(ctx))
+
+		authheader := e.Request().Header.Get("Authorization")
+		pref := "Bearer "
+		if !strings.HasPrefix(authheader, pref) {
+			return echo.ErrForbidden
+		}
+
+		token := authheader[len(pref):]
+
+		exists, err := bgs.lookupAdminToken(token)
+		if err != nil {
+			return err
+		}
+
+		if !exists {
+			return echo.ErrForbidden
+		}
+
+		return next(e)
+	}
+}
+
+type User struct {
+	ID          models.Uid `gorm:"primarykey;index:idx_user_id_active,where:taken_down = false AND tombstoned = false"`
+	CreatedAt   time.Time
+	UpdatedAt   time.Time
+	DeletedAt   gorm.DeletedAt `gorm:"index"`
+	Handle      sql.NullString `gorm:"index"`
+	Did         string         `gorm:"uniqueIndex"`
+	PDS         uint
+	ValidHandle bool `gorm:"default:true"`
+
+	// TakenDown is set to true if the user in question has been taken down.
+	// A user in this state will have all future events related to it dropped
+	// and no data about this user will be served.
+	TakenDown  bool
+	Tombstoned bool
+
+	// UpstreamStatus is the state of the user as reported by the upstream PDS
+	UpstreamStatus string `gorm:"index"`
+
+	lk sync.Mutex
+}
+
+func (u *User) GetDid() string {
+	return u.Did
+}
+
+func (u *User) GetUid() models.Uid {
+	return u.ID
+}
+
+func (u *User) SetTakenDown(v bool) {
+	u.lk.Lock()
+	defer u.lk.Unlock()
+	u.TakenDown = v
+}
+
+func (u *User) GetTakenDown() bool {
+	u.lk.Lock()
+	defer u.lk.Unlock()
+	return u.TakenDown
+}
+
+func (u *User) SetTombstoned(v bool) {
+	u.lk.Lock()
+	defer u.lk.Unlock()
+	u.Tombstoned = v
+}
+
+func (u *User) GetTombstoned() bool {
+	u.lk.Lock()
+	defer u.lk.Unlock()
+	return u.Tombstoned
+}
+
+func (u *User) SetUpstreamStatus(v string) {
+	u.lk.Lock()
+	defer u.lk.Unlock()
+	u.UpstreamStatus = v
+}
+
+func (u *User) GetUpstreamStatus() string {
+	u.lk.Lock()
+	defer u.lk.Unlock()
+	return u.UpstreamStatus
+}
+
+type UserPreviousState struct {
```

Comment:
can this just be merged on to the existing `Users` table? I guess this table has higher churn and it might be good to keep that separate

Reply:
yeah, that's my motivation. I think Postgres works better with the high churn on separate smaller rows from the relatively low-churn data.